crates/cli/src/main.rs

```diff
@@ -1,52 +1,226 @@
-use anyhow::{anyhow, bail, Result};
+use anyhow::{anyhow, bail, Context, Ok, Result};
 use clap::Parser;
 use core::fmt::Write;
-use std::fs;
+use std::{
+    ffi::OsStr,
+    fs,
+    path::{Path, PathBuf},
+};
 use wasmi::{
     core::{Value, ValueType, F32, F64},
+    Engine,
     ExportItemKind,
     Func,
     FuncType,
+    Linker,
+    Module,
     Store,
 };
+use wasmi_wasi::{ambient_authority, Dir, TcpListener, WasiCtx, WasiCtxBuilder};
+
+fn parse_env(s: &str) -> Result<(String, String)> {
+    s.split_once('=')
+        .map(|(s1, s2)| (s1.to_owned(), s2.to_owned()))
+        .ok_or_else(|| anyhow::anyhow!("expects '='-separated pair"))
+}
 
 /// Simple program to greet a person
 #[derive(Parser, Debug)]
 #[clap(author, version, about, long_about = None)]
 struct Args {
-    /// The WebAssembly file to execute.
-    #[clap(value_parser)]
-    wasm_file: String,
+    /// The host directory to pre-open for the `guest` to use.
+    #[clap(long = "dir", number_of_values = 1)]
+    dirs: Vec<String>,
+
+    /// The socket address provided to the module. Allows it to perform socket-related `WASI` ops.
+    #[clap(long = "tcplisten", number_of_values = 1)]
+    tcplisten: Vec<String>,
```

Comment:
```suggestion
    #[clap(action = clap::ArgAction::Append)]
    tcplisten: SocketAddr,
```

Reply:
`SocketAddr` refer's to Rust's built-in type: https://doc.rust-lang.org/std/net/enum.SocketAddr.html#impl-FromStr-for-SocketAddr
Note that this requires an implementation of [`clap::TypeValueParser`](https://docs.rs/clap/latest/clap/builder/trait.TypedValueParser.html) for `SocketAddr` because for some reason it does not seem as if `clap` already does this for us which I think it should.
Anyhow the implementation should be fairly similar to both:
- https://docs.rs/clap/latest/src/clap/builder/value_parser.rs.html#950-978
- https://docs.rs/clap/latest/src/clap/builder/value_parser.rs.html#860-886

Reply:
For this you need to define a tiny wrapper type around `SocketAddr` and implement the trait for that type due to orphan rules.

Reply:
It works without implementing `clap::TypeValueParser` for `SocketAddr`. They must've already implemented it.